### DISCUSSION TITLE
RemoteGPIO enhancements

### DISCUSSION THREAD

**Discussion by drtinaz** [Lucifer06/RemoteGPIO#31] [Ideas]
I didn't realize I was replying to your announcement when I should have started a new thread. Anyways, here goes.

Hello Frederic. Since you went through the effort to open this discussion board I thought I might as well start using it. This regarding our previous discussion concerning high system resources usage and possible freeze-up and reboot when using three modules.

I have made pretty good progress in my limited time testing. I have been able to reduce the overall CPU usage by 65 - 70% by making a few changes to the service and driver. I am going to continue testing. I would also like to make a few changes to the RemoteGPIO settings menus such as adding the ability to enable/disable reading external relay state and digital inputs for each module individually. I would also like to add a menu option to restart the rgpio service from within the menu for the following reason. I removed the dynamic updating of number of units and number of relays from the service loop as this was the highest usage of overall CPU usage and system resources, and added a one liner to run S90rgpio_pins.sh at startup before the loop. This alone results in a 50% reduction in system resource usage. I have tested this to work properly upon system reboot and firmware upgrade. What do you think? I don't want to spend a lot of time on changing/adding menu options if you are not in agreement.

**Comment by drtinaz** [comment]
I've made pretty good progress so far. I have added an "Additional Options" menu for each module, which contains toggles to enable/disable reading relay state and digital inputs. I think that the "Module Reboot" should also be moved into this sub menu. I have also added an option to "Restart RemoteGPIO" on the main RemoteGPIO settings page. This makes it possible to make any changes to the configuration without having to reboot the GX device.

One issue I have run into is that when digital inputs are disabled, the string value for the latency is not being read properly after a reboot. A restart of the driver fixes it but I can't see anything that would cause it. I might try adding a sleep delay at the top of the driver to give more time for the settings to load but I can't see that being the issue. For now I have just set the latency to a static 0.2 until I get through the testing phase. If you want to have a test run the GitHub is as follows:
packagename:RemoteGPIO username:drtinaz branch:dev3

I know you are currently travelling, as am I.

**Comment by Lucifer06** [comment]
Hello Thomas we are in agreement, I don't see any benefits to have dynamic i/O configuration.I'm in LA flying back home today...Cheers,FreD. Excuse possible typo: swallowed by my iPhoneLe 18 avr. 2024 à 07:45, Derrick ***@***.***> a écrit :﻿
I didn't realize I was replying to your announcement when I should have started a new thread. Anyways, here goes.
Hello Frederic. Since you went through the effort to open this discussion board I thought I might as well start using it. This regarding our previous discussion concerning high system resources usage and possible freeze-up and reboot when using three modules.
I have made pretty good progress in my limited time testing. I have been able to reduce the overall CPU usage by 65 - 70% by making a few changes to the service and driver. I am going to continue testing. I would also like to make a few changes to the RemoteGPIO settings menus such as adding the ability to enable/disable reading external relay state and digital inputs for each module individually. I would also like to add a menu option to restart the rgpio service from within the menu for the following reason. I removed the dynamic updating of number of units and number of relays from the service loop as this was the highest usage of overall CPU usage and system resources, and added a one liner to run S90rgpio_pins.sh at startup before the loop. This alone results in a 50% reduction in system resource usage. I have tested this to work properly upon system reboot and firmware upgrade. What do you think? I don't want to spend a lot of time on changing/adding menu options if you are not in agreement.

**Reply by cbelcher** [reply]
Derrick, I'm Craig pleasure to make your acquaintance.

Not a GitHub expert by any means, do we start a new branch for dev team?  My work has solved the Digital Input issue on the Pi and Derrick's work, is obviously a huge issue.  Finally have my Cerbro GX MK2 and see just how horribly under powered it is, compared to the Pi 4 Model B.  After installing on the Cerbo, it just cripples it.  Run's like a clock on the Pi.  Still working on the dbus-digitalinput.py script tracing couple times and hour.  Good chance Derrick's work will resolve this.  From the sound of it, he's addressing the very issue that's throwing the the exception on the Pi.  Going to pull his repo as I type.  Let's pull this team together and make a rock solid solution for all to use and enjoy.

**Reply by cbelcher** [reply]
Fredrick wrote Thomas, is that Derrick?  Now I'm all confused :)  We'll get the names all worked out.  Cheers

**Reply by drtinaz** [reply]
Derrick is my first name, Thomas is my last name. Either one is fine. I get that all the time since both are common as first names.

**Comment by cbelcher** [comment]
Settled I’m calling you Derrick. 

Do we know if this works on the Ekrano?  Gut says yes since it comes 2 Digital Inputs, and passes start-digitalinputs.sh test, that was killing it on the Pi.  I’ll be getting one here shortly, for no other reason that this Cerbo GX MK2 is such a CPU bound mess.  There’s no way it’s going in my new house, not a chance.

I didn’t get a chance to pull your work last night, very interested to see what you’ve been up to.  If by chance it doesn't solve my exception issue on the Pi, it'll at least point me in the right direction. 

Cheers,

**Reply by drtinaz** [reply]
I'm still travelling today, I will be home this evening and then I will have more time over the next few days for further testing and finishing up on some enhancements and refining the services. Should make it run a little better on the GX platform.

It should work on the ekrano as they are essentially the same OS and platform. Difference would be that there would be no digital input 3 and 4 since the ekrano only has 2 and RemoteGPIO installs 5 and up according to the configuration. It would probably be possible however to modify the S90 script to accommodate the different numbering schemes according to what device is being used. It could either run a check to see which platform it is running on, or add a setting to select the platform. What do you think @Lucifer06 ?

**Reply by drtinaz** [reply]
BTW, @cbelcher if you think it's sluggish on your mk2, you should see it run on my production system. I have parallel MP2, 2 mppt's, 2 smart shunts, 7 tank sensors, 6 temp sensors, 3 relay modules, and bms all running on the mk1 with guimods. Talk about sluggish. If I run the modules through tcp in node red they are really snappy, but I have been working on trying to 'trim the fat' on the Dbus version as I would really rather run it natively without depending on node red. I find that the service and the driver use minimal CPU but the other resources being used by the driver and service is what really taxes the CPU.

**Comment by kwindrem** [comment]
Restarting the service to handle big changes is probably OK but should be automated. Your code can trigger a restart when it detects changes to the associated dbus parameters. The simplest way to restart is to exit from mainLoop. The script will then exit and the service will be restarted.

I'm not sure I see the need to read relay state values from the relay box. This should be one-way communications: GX should always set the relay value and there should be no other systems controlling the same relay.

One thing I've done to reduce execution time is to keep local values for dbus parameters. If dbus parameters can change from outside the program (eg values set by the GUI), I use the dbus change handler to update the local copy.

If you are setting a dbus Setting parameter often, you may need to slow that down to avoid time in the non-volatile mechanism. As an example the generator code does this. It keeps a local variable that's updated frequently, then the dbus Setting is updated from that every 30 seconds (or something like that). If you need a dbus value to be more up to date, create a parameter in your service and use that everywhere. The Settings value is only there for non-volatile storage. This is what is done for the relay states so they are restored after a restart. But the value in ...system is the working one.

CCGX is far more limited in processing power than Cerbo. So you either need to address CPU usage on CCGX or specifically block install of your package on that platform. Yes, RPI 4 has more processing power than Cerbo so you need to test CPU utilization on that. Keep in mind gui-v2 is also gobbling up CPU bandwidth so I'd check with that running locally also. A remote gui-v2 actually runs on the platform hosting the display, not on the GX device so you need to run gui-v2 on the local display.

You might also look into multi-threading your service and scheduling different activities at different frequencies. PackageManager runs several threads, many of which wait on something placed in the thread's queue. Separate threads can reduce the tasks your main loop has to do every second.

I use a timer to measure execution times so I can identify where to focus efforts to optimize performance. PackageManager's main loop has such a timer (commented out) that you can use as a template.

**Reply by drtinaz** [reply]
I have already removed a lot of the code that restarts the service automatically on Dbus changes and now the service only restarts by a manual selection in the settings. Configuration changes no longer restart the service.

As far as reading relay state, there may be some situations (I have several) where the relay may be switched from an external source such as a wall switch, and the gui has no way of knowing the state has changed. 

Some major changes have been made on my dev branch.
Packagename:RemoteGPIO username:drtinaz branch:dev3

Still a work in progress.

**Comment by cbelcher** [comment]
This was your last email I think I see others.

Man been on a network down all day.  Entire Florida electric grid, they can monitor all the substations.  Boy talk about pressure.

A work in progress, for sure.  But I’m having fun.

**Comment by drtinaz** [comment]
I have made enhancements on the dev3 branch. Now ready for beta testing.

**Comment by drtinaz** [comment]
to all, I have merged the dev3 into the dev branch. If there are no issues and @Lucifer06 approves of the changes I will push to the main branch. For now you can give it a try @ package name:RemoteGPIO username:drtinaz branch:dev

@Lucifer06 I temporarily removed the user setting for latency as I was having issues with the string not loading after a reboot. I have set sleep 0.1 which seems to be fairly snappy. Any more than that and the delay between switching on the gui and the relay changing state becomes noticable. The difference in CPU usage between 0.1 latency and higher seems to be less than 1% so I don't see the need to add the user setting back into the menu. What do you think? I have discovered what was causing the latency not to load properly so I can add the setting back pretty easily if you decide that would be beneficial.

And lastly, the changes I have made here have resulted in an overall CPU usage reduction of up to 75% for the package. I have been running it on my very busy cerbo and have had no freeze ups or reboots. The system can get a little sluggish from time to time, but that's most likely just because of the number of devices I have connected, and some extensive node red flows.

Feedback is appreciated.

**Comment by cbelcher** [comment]
Derrick what’s your GMT offset.  In Tampa (this week it’s) -5.   They’ve tried to kill it for decades this insane practice of changing the clocks twice a year, it’s maddening.

I’m guessing Fredrick must be +1’sh in good old France?

Was meaning to ask,  you doing your testing on a Cerbo?   I’m only 4’ish months from, even knowing this company existed.   Now that’s a rookie, for you 😊

Cheers can’t wait to pull and test.

**Reply by drtinaz** [reply]
Currently pacific time. We don't do dst here in Arizona.

**Comment by cbelcher** [comment]
Ya I know!  It’s the worst.  Just moved from the South bay in LA.  Lord I miss it, had to come back East to take care of mom.

Not having much experience with GitHub is the best way to communicate, or is there a chat system build into their desktop?  Haven’t had a chance to look.

**Comment by cbelcher** [comment]
My changes to the S90rgpio_pins.sh script updates. Has passed on both the Cerbo and RPi, 3 & 4’s.  Next is to test the RPi with the Waveshare (non-FD) variant and see if it flies.
Need to order up a FD Hat and make sure that works as well.

It’s just a suggestion, how to deal with the issue.  I made a call to new subroutine get_platform_setting that just reads Fredricks packageManager/Platform value and parses it for the 3rd literal. Kicks back PI for both the Pi3 and 4 variants.  The I go to town, dealing with the changes needed to get it running on the PI, with some case statements.

Am I using that term correctly “literal”?  Been more than a couple decades, since computer science programming classes. 😊 It just came out, when typing.

I’ll try to get it on GitHub for you guys to have a look, don’t laugh 😊  I’m a hardware engineer.  As you both know, most hardware isn’t worth the sand it was made from without your skills.  God bless you both.

Happy Sunday.

**Comment by Lucifer06** [comment]
Hello Guys sorry for the radio silence, I had to travel back to back for couple past weeks. Now I need to digest the 9 hours jet lag..
Hope to be able to test validate all the changes you guys have made.
Are Craig's changes for rPi support already in Derrick's dev distribution? 

Regarding the Latency setting: I think it's fine we can remove this setting. That setting was a dirty workaround we make some optimisation...

For Numbering the Digital Inputs and also the relays: Given the various platforms (rPI, Ekrano, Cerbo) that has from 0 unto 4 DI and Zero or 2 Relays, we need to discuss if we always want to have no discontinuity with the numbering (will require some more work ) or if RemoteGPIO will always start with the same numbering regardless of host platform. I propose we discuss this particular point in a separate discussion.

**Reply by drtinaz** [reply]
Craig's changes are not yet incorporated. I propose we look at finishing the enhancements made in the dev branch, and then Craig should issue a PR with the pi changes.

**Reply by Lucifer06** [reply]
OK then will test package name:RemoteGPIO username:drtinaz branch:dev

**Comment by cbelcher** [comment]
Oh yes, been mulling over that very topic since I started.  I can’t think of a solution.

Didn’t really dig into it just because I know you said you had them all hard coded.  Love to bounce ideas off and see if we can come up with something really slick.

**Comment by cbelcher** [comment]
What’s a PR?  Pull request, I’m learning 😊

**Reply by drtinaz** [reply]
If you haven't already, create a fork of Frederic's RemoteGPIO (main branch as main) and then on your GitHub you can make and test updates/revisions. Once you have verified all is good then you can create a PR back the the upstream main.

**Comment by Lucifer06** [comment]
@drtinaz I'm testing v3.2.17dev
I did the setup from the GUI.
I got the GUI restarted automatically, and when reconnecting in package manager, I had the message to restart the GUI. And I never got offered to reboot Venus.
Any changes that would make the GUI to restart automatically without asking? And no more reboot required?
May be it's because I already had one unit configured. May be we need to pre-configure to Zero unit for the setup?

I like the idea of the sub-menu with additional options and good idea to remind about Addr setting depending of Unit#
I'm not sure you need to configure the TCP-Client on the Dingtian GUI. TCP-Server configured as "Modbus-RTU over TCP" should be enough.

**Reply by kwindrem** [reply]
Unless the package's setup script overrides GuiRestart and/or SystemReboot, HelperResources will determine what needs to be done after a package install/uninstall finishes. If GUI V1 files are modified, a GUI restart is indicated. If some other files are modified (like gpio_list) that are read only during system boot, then a reboot is required.

So if you reinstall a package and there have been no changes to the package it is likely no restarts or reboot will be required.

**Reply by drtinaz** [reply]
There are no overrides in the setup script so helper resources is handling restarts/reboots. The S90 script modifies the relay pins which in turn changes what is or is not being displayed on the relays overview, which is why I added the gui restart to the S90 script. The S90 script always runs at system reboot, or service restart after a configuration change. @kwindrem would you suggest handling this a different way?

**Reply by kwindrem** [reply]
Any files modified in /etc/udev/rules.d/* from the setup script will require a system reboot so I don't think the s90 script should not do its own reboot. That might result in a reboot loop.

**Reply by drtinaz** [reply]
The S90 script only does a gui restart, no system reboot.

**Reply by kwindrem** [reply]
udev rules are only used at boot (or possibly other service restarts but that might be difficult to determine which ones). I can't see how modifying one of these scripts would affect the GUI. I looked at the code (main) and don't see where you are restarting the gui in S90rgpio_pins.sh but that might be something you haven't merged into main yet.

The reason SetupHelper manages system reboots and GUI and other service restarts is so that all packages can be installed/uninstalled prior to doing the reboot/restart. Otherwise, the GUI might reset with every package install.

**Reply by drtinaz** [reply]
> udev rules are only used at boot (or possibly other service restarts but that might be difficult to determine which ones). I can't see how modifying one of these scripts would affect the GUI. I looked at the code (main) and don't see where you are restarting the gui in S90rgpio_pins.sh but that might be something you haven't merged into main yet.
> 
> The reason SetupHelper manages system reboots and GUI and other service restarts is so that all packages can be installed/uninstalled prior to doing the reboot/restart. Otherwise, the GUI might reset with every package install.

The dev branch has not yet been merged into main. Dev is on my GitHub drtinaz.

**Reply by kwindrem** [reply]
Oops, I was thinking about udev not init. SetupHelper currently does not check init files for changes since these should only be called at system startup/shutdown. A file change here should trigger a reboot and I will add this to my checks. However, simply rerunning your replacement S90 script would not trigger a reboot or any other restart. You would need to do that somewhere in your code if needed.

I'd like to look into the issue with the relay overview not refreshing when the S90 script is run after boot. What's on the main branch doesn't see to affect the relay overview if I run it manually. I do have code looking for a change in value of show/hide each relay but may need to look for other dbus value changes also.

I can't seem to find the Dev branch and from other discussions here I'm guessing it is private.

**Reply by drtinaz** [reply]
> Oops, I was thinking about udev not init. SetupHelper currently does not check init files for changes since these should only be called at system startup/shutdown. A file change here should trigger a reboot and I will add this to my checks. However, simply rerunning your replacement S90 script would not trigger a reboot or any other restart. You would need to do that somewhere in your code if needed.
> 
> I'd like to look into the issue with the relay overview not refreshing when the S90 script is run after boot. What's on the main branch doesn't see to affect the relay overview if I run it manually. I do have code looking for a change in value of show/hide each relay but may need to look for other dbus value changes also.
> 
> I can't seem to find the Dev branch and from other discussions here I'm guessing it is private.
> 

Not private, but it is on my GitHub and not on Frederic's which is probably why you were not able to find it.

Packagename:RemoteGPIO
Username:drtinaz
Branch:dev

Lowercase d on the branch.

**Reply by kwindrem** [reply]
I just tried the dtrinaz/dev branch and commented out the gui restart line in the S90 script and ran it manually. The relay overview refreshed properly without a gui restart. This test was without any actual relay boxes attached. The current state fields went to ?? briefly but came back within a couple of seconds.

**Reply by kwindrem** [reply]
FYI: the GUI restart code in s90 will not work with RPI at the moment. RPI is still using /service/gui for the GUI, not service/start-gui like is used on other platforms. This will eventually change (probably) when Victron implements gui-v2 on RPI.

There's code in PackagManager (towards the end) that you can look at so you can restart the correct service.

**Reply by drtinaz** [reply]
Try toggling the button to disable/enable RemoteGPIO from the settings menu?

I got the blank relay overview on both my test system and production systems. Also try changing the number of relays for the module?

**Reply by kwindrem** [reply]
GuiMods v10.22 should fix the Relay Overview refresh issues. It rebuilds the relay list whenever the number of relays reported by systemCalc changes or when any of the show/hide relays change.

So changing the number of relays per module or the number of modules will rebuild the relay list shown on the Relay Overview.

I don't think you need to restart the GUI when running the S90 script.

**Reply by drtinaz** [reply]
Ok I will take it out.

**Comment by drtinaz** [comment]
Reconfiguring to zero for setup results in having to reset number of units after any package update as well as Venus os firmware update, which is why I removed it. If a user has auto install enabled in package manager and there is an update this would result in RemoteGPIO needing to be reconfigured before the relay modules will function again.This makes for a more seamless update.

The gui Auto restarts anytime S90rgpio_pins.sh is ran. Without this, if the gui is on the relay overview page when S90 rgpio is ran it results in a blank overview page which can only be fixed by navigating to the settings page and then back to the overview page. S90rgpio is only ran at system reboot or package install (other than a manual restart in the settings menu). I think this is a fair tradeoff. What do you think?

Venus reboot is not required because you already had the package installed and configured. A fresh install or uninstall will still reboot Venus automatically.

You are correct, only the server needs to be set to rtu over TCP so that language can be changed.

**Comment by Lucifer06** [comment]
Ok Let me think and deep dive on the code...

**Reply by drtinaz** [reply]
Just released v3.2.18dev.
Fixes TCP setup instructions.

**Reply by drtinaz** [reply]
Let me figure out how to grant you access to my dev branch. I think that would make things easier for you.

**Reply by drtinaz** [reply]
@Lucifer06 you should now have an invitation to colab on the dev branch

**Comment by drtinaz** [comment]
@Lucifer06 v3.2.19dev just released. This one removes gui restart from S90 script. @kwindrem has resolved the blank relay overview page issue from within guimods.

**Comment by cbelcher** [comment]
I’ll get my S90 so you guys can use it as a base to get your RPI users back online.

**Reply by kwindrem** [reply]
With the changes I just made in GuiMods, the GUI restart code in S90 is no longer required.

**Comment by cbelcher** [comment]
Just finding some time to get back on this.

Derrick, just about to test again to see if I’m not crazy.  Starting on line 82 of the S90, where you set all the DI’s types to 0. Could have sworn this removed all my DI from the Device List.  Don’t recall every testing this on my GX.  But I know on my latest running version I’ve got them all commented out on the RPI case.

Testing this right after I send this.

Then I’m off to figure out how to fork your project and publish it 😊

**Comment by cbelcher** [comment]
Confirmed.  Reboot of the RPI with those line uncommented, DI’s gone from Device List.

I’ve just included screenshots of Device List before and after and a shot of the S90 lines commented and uncommented.  Need to try it next on the Cerbo.

**Reply by drtinaz** [reply]
@cbelcher give me a few minutes, I want to make a few changes. I see no reason for the inputs to be set to disabled in the settings when there is a configuration change. The settings should always survive no matter what. Simply removing the digital input links should be enough to remove them from the device list when there is a configuration change. I don't want the user to have to reconfigure the digital inputs any time there is a configuration change or an update.

Stay tuned, update coming momentarily.

**Comment by cbelcher** [comment]
Take your time.

Just kind of makes what little sense I have; if you yank its Type value it goes away?  And it does, just commented them out again and the DI’s survive a reboot.

Going to test on the GX, they must behave differently, or this would be on your radar.

**Reply by drtinaz** [reply]
It will act the same on the gx. I missed it because I am not using the gx to read inputs.

v3.2.21dev just released and should fix this issue.

**Comment by cbelcher** [comment]
Too cool.

**Reply by drtinaz** [reply]
I do not have a pi to test on so your feedback is appreciated. Do the digital inputs work as expected on your pi with the last update?

**Comment by cbelcher** [comment]
Pulling Repo as I type.

**Comment by cbelcher** [comment]
We are talking the main branch, still just see 3.2.2.

**Reply by drtinaz** [reply]
No. This discussion thread is for the dev branch.
packagename:RemoteGPIO
username:drtinaz
branch:dev

If you want to try the dev package you will need to uninstall and remove the current package from package manager and add the dev package using the inactive packages menu. 

There are many changes and enhancements to the dev branch

**Comment by cbelcher** [comment]
Do I need rights to this branch?  From GitHub Desktop the only branches visible to me are: main & origin/old.

**Reply by drtinaz** [reply]
It sounds like you are on Frederic's GitHub. You will not find it there. It is on my GitHub.
https://github.com/drtinaz/RemoteGPIO/tree/dev

**Comment by cbelcher** [comment]
This is a shot of my GitHub Desktop?  Let’s see if look different in web browser of your project.

**Comment by cbelcher** [comment]
Slight naming difference with 2nd branch, but close enough.

**Reply by drtinaz** [reply]
@cbelcher if you are posting images I am not seeing any of them. You might try responding directly on GitHub instead of thru your email.

**Comment by cbelcher** [comment]
Roger that.

**Comment by cbelcher** [comment]
Let's see if you get my images from here.
GitHub Desktop I can't see the branch, I looks like this.
![Screenshot 2024-04-23 14 15 08](https://github.com/Lucifer06/RemoteGPIO/assets/4452789/d3792aeb-22ba-4c8f-a059-9450483e734a)

From GitHub website like so:
![Screenshot 2024-04-23 14 18 09](https://github.com/Lucifer06/RemoteGPIO/assets/4452789/d810d745-b521-44ae-8bad-efa90ba5a770)

What am I doing wrong?

**Reply by drtinaz** [reply]
You are on Frederic's GitHub. The dev branch is on my GitHub.

https://github.com/drtinaz/RemoteGPIO/tree/dev

**Comment by cbelcher** [comment]
You are correct.  Your link took me to your fork. Thanks man!

**Comment by cbelcher** [comment]
Derrick, thanks for setting me straight.  Off to testing.

**Comment by Lucifer06** [comment]
I'm now after your last v3.2.21dev release.
All looks fine, except an issue with the read relay status (same issue on v3.2.18dev).
In certain conditions, if you hit on/off of multiple relays very quickly, the GUI display the change, but it's possible that one of the relay will not switch, and few seconds later, when the read of the status is triggered, the GUI get refreshed, displaying the correct status of the relay that got missed. This is more easy to reproduce when controlled via RS485.
When Read Relay State is disabled, I can't reproduce the issue.
This must be a race condition, and can't find it...  May be not a big deal as not everybody will enable this feature.

You have done a great job in S90rgpio_pins.sh !!!

I may start pulling all the changes on the main RemoteGPIO :-)

**Reply by kwindrem** [reply]
Undoubtably a race condition. RS485 will probably be slower than TCP. I haven't looked at the code but I would guess that the reading of the external relay state is happening in a separate thread than the sending of the GX state to the external relay. There may also be a delay in the external box reporting the changed state. You would need delay reading the external relay state after you set a relay state from the GX device.

**Reply by Lucifer06** [reply]
Actually this is indeed managed in two separate services. It may require to sync each other.

**Comment by cbelcher** [comment]
Found it, PackageManager.py.  Too cool, I can only imagine the number of hours Kev’s put into this.  Incredible.
